server/server/src/main/kotlin/org/jetbrains/bazel/server/sync/JvmToolchainQuery.kt

```
@@ -41,4 +42,92 @@ object JvmToolchainQuery {
       }
     }
   }
+
+  suspend fun jvmToolchainQueryForTarget(
+    bspInfo: BspInfo,
+    bazelRunner: BazelRunner,
+    workspaceContext: WorkspaceContext,
+    target: Label,
+  ): JvmToolchainInfo {
+    val aqueryExpression = "mnemonic(\"Javac\", $target)"
+    val command =
+      bazelRunner.buildBazelCommand(workspaceContext = workspaceContext, inheritProjectviewOptionsOverride = true) {
+        aquery {
+          options.addAll(listOf("--output=jsonproto", aqueryExpression))
+        }
+      }
+    val aqueryResult =
+      bazelRunner
+        .runBazelCommand(command, logProcessOutput = false, serverPidFuture = null)
+        .waitAndGetResult(ensureAllOutputRead = true)
+    if (aqueryResult.isNotSuccess) throw RuntimeException("Could not query target '$target' for jvm toolchain info")
+
+    // Extract target-specific java_home, toolchain_path, and jvm_opts from aquery
+    val (javaHome, toolchainPath, jvmOpts) = parseJavaHomeAndToolchainPathFromAquery(aqueryResult.stdout)
+
+    return JvmToolchainInfo(
+      java_home = javaHome,
+      toolchain_path = toolchainPath,
+      jvm_opts = jvmOpts,
+    )
+  }
+
+  /**
+   * Extract java_home, toolchain_path, and jvm_opts from aquery result.
+   * Uses -jar as delimiter: everything before -jar are JVM options, everything after are compiler options.
+   */
+  private fun parseJavaHomeAndToolchainPathFromAquery(stdout: String): Triple<String, String, List<String>> {
+    val gson = Gson()
+    val jsonObject = gson.fromJson(stdout, JsonObject::class.java)
+    val actions = jsonObject.getAsJsonArray("actions")
+
+    if (actions.isEmpty) {
+      throw RuntimeException("No Javac actions found in aquery result")
+    }
+
+    val javacAction =
```

this check and the one above may be redundant?

---

I'd keep it, just in case. I'm not sure if that is possible but better that as silent failure